### ISSUE TITLE
Resource Management

### ISSUE THREAD

**Issue by logxen** [ScreepsOCS/screeps.behaviour-action-pattern#679]
~To prepare for labs our creeps need the ability to decide where resources should go (instead of just stuffing everything into the storage with overflow going to the terminal). This system will manage all minerals and compounds but probably not energy and power.~

~`Room.resources.setLoadingAmount(<structure>, <resource>, <amount>)`~

~This command populates a data object that stores what the player wants to happen with a room's resources.~

~`structure` is the lab, container, storage, or terminal to be controlled.~
~`resource` is the resource to set the target amount for.~
~`amount` is the amount of 'resource' that haulers will proceed to attempt to load into the structure.~

~To empty a structure of a resource set the amount to 0.~
~Running this command on a lab will set the loading amount to 0 for all other resources for that lab.~

~Hauler creeps will then attempt to fill structures to the requested loading amounts as part of action.reallocating. Structures will be filled in the following order of priority:~
~Lab -> Container -> Terminal -> Storage (Storage still overflows to the Terminal for auto-sell)~

~By default all structures are assumed to have loading amounts of 0. For the storage Structure any set loading amount is *added* to the savings amount set in parameters. As such, once this is in place everything should work as before until a setLoadingAmount command is actually executed.~

The info in this post is superseded by the comments below.

**Comment by karlthepagan** [comment]
For spawning boosted creeps the requirements are slightly different, but this is good to get us started.

**Comment by logxen** [comment]
Ok ... on further consideration I think it needs two functions. One that is a one-time order and another that is a continuous storage order. This would allow more careful manipulation of the resources. So:

room.placeOrder(<structure>, <resource>, <amount>)
room.setStore(<structure>, <resource>, <amount>)

This will extend the data structure a bit as well, changing `resourceType: targetAmount` to:
```javascript
    resourceType: {
        'orderAmount': orderAmount,
        'orderRemaining': orderRemaining,
        'storeAmount': storeAmount
    }
```

The resourceType.orderRemaining value would start equal to the total orderAmount but every time a hauler dropped off the resource it would subtract the amount from the orderRemaining. In this way the haulers know both how much of a one time order is left to be filled and they can look at the orderAmount plus the storeAmount and thereby would not immediately bring the resource back to storage.

To reset the value every time a hauler takes the resource from storage he should subtract that amount from orderRemaining. This will effectively let the one-time resources be used first.

**Comment by logxen** [comment]
The original data structure I have sketched out up there is flawed ... here is the proper one I have implemented now with arrays where necessary and such:

```javascript
Memory.rooms.resources = {
  lab : [
    { 
      id : containerId,
      orders : [
        {
          type : resourceType
          orderAmount : 0
          orderRemaining : 0
          storeAmount : 0
        },
        ...
      ]
    },
    ...
  ],
  container : [
    ...
  ],
  terminal : [
    ...
  ],
  storage : [
    ...
  ]
}
```

**Comment by logxen** [comment]
Just implemented lab orders!

Game.rooms[<roomName>].placeReactionOrder(<labId>, <resourceId>, <amount>)

The lab will search for 2 nearby idle labs to use as sources and will automatically place resource orders for them.

This command does not set a store amount on the master lab so usually haulers would immediately start carrying the produced compound to storage.

**Comment by logxen** [comment]
Ok ... I made the lab system recursive because it was pretty simple to do and now it can theoretically do tier 3 reactions with a single command if all is setup right. (I don't have an RCL8 room in my test environment yet though)

Also here's the new command for inter-room trade orders:

```javascript
Game.rooms[<roomName>].placeRoomOrder(<orderId>, <resourceType>, <amount>)
```

Currently the only situation where this gets called is when a hauler running reallocation task cannot find the resources it needs for a lab in the room. It will then place a room order for the needed goods. This should both enable fully automated lab reactions as well as aid in moving compounds to where they are needed for boosting.

**Comment by whardeman** [comment]
@logxen I have a couple RC8's, so I'm happy to test what you've done so far. Sounds like great work.

I may not have time until Monday, though.

**Comment by logxen** [comment]
@whardeman @Almaravarion yes, please ... hit me up on slack.

Inter-room trading is now functioning in the branch! Woot! You have to turn on ROOM_TRADING in parameter.js or they won't trade between rooms on their own. With this we nominally have fully automated production.

There are of course a few bugs that need to get tracked down here and there ... there are SO MANY potential behaviors just from moving resources around the same room in creep.action.reallocation that it's really hard to actually test all possibilities. Most of the basic stuff I have tried seems to be working though.

**Comment by logxen** [comment]
The main branch has been merged and I have created a wiki page so I'm closing this thread
